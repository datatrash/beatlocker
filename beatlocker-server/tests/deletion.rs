use crate::test_utils::{copy_recursively, TestClient};
use axum::http::StatusCode;
use beatlocker_server::*;
use chrono::{DateTime, Utc};
use std::fs::{create_dir_all, remove_dir_all};
use std::path::PathBuf;
use std::sync::Arc;

#[path = "test_utils/mod.rs"]
mod test_utils;

use test_utils::*;

struct DirDeleter(PathBuf);

impl Drop for DirDeleter {
    fn drop(&mut self) {
        remove_dir_all(&self.0).unwrap();
    }
}

#[tokio::test]
async fn deletion_test() -> AppResult<()> {
    let temp_path = PathBuf::from("./deletion_test_data");
    create_dir_all(&temp_path)?;
    let _deleter = DirDeleter(temp_path.clone());
    copy_recursively("tests/data", &temp_path)?;

    enable_default_tracing();
    let options = ServerOptions {
        path: temp_path.clone(),
        database: DatabaseOptions {
            path: Some(PathBuf::from(".")),
            in_memory: true,
        },
        now_provider: Arc::new(Box::new(|| {
            DateTime::parse_from_rfc3339("2020-02-02T00:00:00Z")
                .unwrap()
                .with_timezone(&Utc)
        })),
        ..Default::default()
    };
    let app = App::new(options).await?;
    let client = TestClient::new(app.app.clone());

    app.task_manager
        .send(app.import_all_folders().await?)
        .await?;

    std::fs::remove_file(temp_path.join("Richard Bona/Richard Bona - Ba Senge.ogg"))?;

    app.task_manager
        .send(app.remove_deleted_files().await?)
        .await?;

    let res = client
        .get(&format!("/rest/getArtist?f=json&id={RICHARD_BONA_UUID}"))
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!(
        "getArtistAfterDeletingBaSenge.json",
        res.json::<serde_json::Value>().await
    );

    remove_dir_all(temp_path.join("Richard Bona"))?;
    app.task_manager
        .send(app.remove_deleted_files().await?)
        .await?;

    let res = client.get("/rest/getAlbumList?f=json").send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!(
        "getAlbumListAfterDeletingAlbum.json",
        res.json::<serde_json::Value>().await
    );
    let res = client.get("/rest/getArtists?f=json").send().await;
    assert_eq!(res.status(), StatusCode::OK);
    insta::assert_json_snapshot!(
        "getArtistsAfterDeletingAlbum.json",
        res.json::<serde_json::Value>().await
    );

    Ok(())
}
