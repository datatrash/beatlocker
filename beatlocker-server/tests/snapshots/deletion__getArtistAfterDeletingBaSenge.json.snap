---
source: beatlocker-server/tests/deletion.rs
expression: "res.json::<serde_json::Value>().await"
---
{
  "subsonic-response": {
    "artist": {
      "album": [
        {
          "artist": "Richard Bona",
          "artistId": "d094e9f8-a8e2-1737-0cfc-c4b24ab0aedf",
          "duration": 27,
          "id": "68bc272d-d36b-9191-b815-02627be8ea65",
          "name": "Tiki",
          "songCount": 1,
          "title": "Tiki"
        }
      ],
      "albumCount": 1,
      "id": "d094e9f8-a8e2-1737-0cfc-c4b24ab0aedf",
      "name": "Richard Bona",
      "song": [
        {
          "album": "Tiki",
          "albumId": "68bc272d-d36b-9191-b815-02627be8ea65",
          "artist": "Richard Bona",
          "artistId": "d094e9f8-a8e2-1737-0cfc-c4b24ab0aedf",
          "contentType": "audio/mp3",
          "created": "2020-02-02T00:00:00Z",
          "duration": 27,
          "genre": "World Music",
          "id": "53c42459-31c6-a4d6-bb12-ac6d1d2dfa23",
          "isDir": false,
          "isVideo": false,
          "parent": "322adc6c-e179-89e7-2e78-8dd857d93e3d",
          "size": 765952,
          "suffix": "mp3",
          "title": "Akwa Samba Yaya",
          "track": 2,
          "year": 2021
        }
      ]
    },
    "serverVersion": "unknown",
    "status": "ok",
    "type": "beatlocker",
    "version": "1.16.1"
  }
}
